src/osemgrep/reporting/Sarif_output.ml

```diff
@@ -0,0 +1,357 @@
+            ("precision", `String "very-high");
+            ("tags", `List (List_.map (fun s -> `String s) tags));
+          ]
+          @ security_severity
+        in
+        let nudge_base =
+          "💎 Enable cross-file analysis and Pro rules for free at"
+        and nudge_url = "sg.run/pro" in
+        let nudge_plaintext = spf "\n%s %s" nudge_base nudge_url
+        and nudge_md =
+          spf "\n\n#### %s <a href='https://%s'>%s</a>" nudge_base nudge_url
+            nudge_url
+        in
+        let text_suffix = if hide_nudge then "" else nudge_plaintext in
+        let markdown_interstitial = if hide_nudge then "" else nudge_md in
+        let references =
+          Option.to_list
+            (Option.map (fun s -> spf "[Semgrep Rule](%s)" s) source)
+        in
+        let other_references =
+          match JSON.member "references" metadata with
+          | Some (JSON.String s) -> [ spf "[%s](%s)" s s ]
+          | Some (JSON.Array xs) ->
+              List_.map
+                (function
+                  | JSON.String s -> spf "[%s](%s)" s s
+                  | non_string -> JSON.string_of_json non_string)
+                xs
+          | Some _
+          | None ->
+              []
+        in
+        let references_joined =
+          List_.map (fun s -> spf " - %s\n" s) (references @ other_references)
+        in
+        let references_markdown =
+          match references_joined with
+          | [] -> ""
+          | xs -> "\n\n<b>References:</b>\n" ^ String.concat "" xs
+        in
+        `Assoc
+          ([
+             ("id", `String (Rule_ID.to_string rule_id));
+             ("name", `String (Rule_ID.to_string rule_id));
+             ("shortDescription", `Assoc [ ("text", `String short_description) ]);
+             ("fullDescription", `Assoc [ ("text", `String rule.message) ]);
+             ( "defaultConfiguration",
+               `Assoc
+                 [
+                   ("level", `String (sarif_severity_of_severity rule.severity));
+                 ] );
+             ( "help",
+               `Assoc
+                 [
+                   ("text", `String (rule_help_text ^ text_suffix));
+                   ( "markdown",
+                     `String
+                       (rule_help_text ^ markdown_interstitial
+                      ^ references_markdown) );
+                 ] );
+             ("properties", `Assoc properties);
+           ]
+          @ rule_url))
+      rules
+  in
+  List.of_seq rules
+
+let fixed_lines (cli_match : OutT.cli_match) fix =
+  let lines = String.split_on_char '\n' cli_match.extra.lines in
+  match (lines, List.rev lines) with
+  | line :: _, last_line :: _ ->
+      let first_line_part = Str.first_chars line (cli_match.start.col - 1)
+      and last_line_part =
+        Str.string_after last_line (cli_match.end_.col - 1)
+      in
+      String.split_on_char '\n' (first_line_part ^ fix ^ last_line_part)
+  | [], _
+  | _, [] ->
+      assert false
```

Comment:
it is unclear what to do in this case -> return an empty list? can it happen? is the assert false fine?

Reply:
assert false is fine if we're sure this can never happen (because OCaml exhaustive check has limitations). If you're not sure, better failwith maybe

Reply:
done in [3bf4518](https://github.com/semgrep/semgrep/pull/9608/commits/3bf45181dd39089747601288efc06bae41cb1fe0)